pkg/pool-utils/contracts/test/MockReentrancyPool.sol

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/pool-weighted/WeightedPoolUserData.sol";
+
+import "../BaseGeneralPool.sol";
+import "../lib/VaultReentrancyLib.sol";
+
+contract MockReentrancyPool is BaseGeneralPool {
+    using BasePoolUserData for bytes;
+    using WeightedPoolUserData for bytes;
+
+    uint256 private immutable _totalTokens;
+
+    event ProtectedFunctionCalled();
+
+    constructor(
+        IVault vault,
+        IVault.PoolSpecialization specialization,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        address[] memory assetManagers,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        address owner
+    )
+        BasePool(
+            vault,
+            specialization,
+            name,
+            symbol,
+            tokens,
+            assetManagers,
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        _totalTokens = tokens.length;
+    }
+
+    // Define a public function protected by the library. Works when called externally; reverts if called
+    // during a Vault operation (join/swap/exit).
+    function protectedFunction() public {
+        VaultReentrancyLib.ensureNotInVaultContext(getVault());
+
+        emit ProtectedFunctionCalled();
+    }
+  
+    // Vault hooks
+
+    // Testing all the hooks is technically overkill: if it works in one Vault context, it works in all.
+    // We mock up an entire pool with all the hooks defined mainly for documentation purposes: in a real pool,
+    // these are all the places you need to protect.
```

Comment:
I find this comment a bit confusing, since it's not the hooks that need to be protected (protecting them will cause basically any valid operation to revert, as the test shows).

I'd propose rephrasing it a bit, saying that these are all the hooks that are called within a vault context, callers should be protected against calling unsafe functions within that context.